src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java

```diff
@@ -68,7 +70,21 @@ private record LinkRequest(FunctionDescriptor descriptor, LinkerOptions options)
     private final SoftReferenceCache<LinkRequest, UpcallStubFactory> UPCALL_CACHE = new SoftReferenceCache<>();
 
     @Override
-    public MethodHandle downcallHandle(FunctionDescriptor function, Option... options) {
+    @CallerSensitive
+    public final MethodHandle downcallHandle(MemorySegment symbol, FunctionDescriptor function, Option... options) {
+        downcallHandle0(Reflection.getCallerClass(), function, options);
+        SharedUtils.checkSymbol(symbol);
+        return downcallHandle(function, options).bindTo(symbol);
+    }
+
+    @Override
+    @CallerSensitive
+    public final MethodHandle downcallHandle(FunctionDescriptor function, Option... options) {
+        return downcallHandle0(Reflection.getCallerClass(), function, options);
+    }
+
+    private final MethodHandle downcallHandle0(Class<?> callerClass, FunctionDescriptor function, Option... options) {
```

Comment:
This method is different from the PR in the panama repo. Here I wanted to try and perform the restricted method check only once.